### PR TITLE
Finished applying fixes for jbmc for assignment consistency

### PR DESCRIPTION
**Pull request by thk123**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message). TODO
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/remove_virtual_functions.cpp
Comment (tautschnig) [inline src/goto-programs/remove_virtual_functions.cpp]: This seems pointless - given the above `type_eq` the difference could now only be tag/symbol vs non-tag/symbol types.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: This PR will need per-commit clang-format cleanup.

File: src/goto-programs/remove_virtual_functions.cpp
Comment (tautschnig) [inline src/goto-programs/remove_virtual_functions.cpp:142]: Ah, please squash this into the commit where I questioned its utility.

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
Oops, missed the do-not-review label.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: documentation here would help

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: :pick: fieldref is not a good name

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:658]: Is fieldref assumed to be a `member_exprt`? It would be better to make that explicit.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:670]: :pick: I would prefer if the loop was expressed as `while(component.is_nil())` rather than `while(1)`

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:681]: :bulb: ideally there should be a function giving the parent class from a java class type, that would make this piece of code clearer.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:675]: :pick: Looks like `resolve_inherited_componentt` can be adapted to do this search

Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:675]: That said - this is just moving code from the typecheck so will clean up in a follow up PR/commit

**Comment by thk123** [comment]
So I couldn't reproduce the failure locally, seemingly because `ctest` does not run the `symex-driven-lazy-loading-expected-failure` target correctly (finds 0 tests to run). I fixed this by removing `-s;symex-driven-loading` from the `CMakeLists.txt` in the relevant regression folder and now get the tests. 

I don't understand this change, so if someone does and wants to PR it, that would have my approval. Otherwise, I will PR it after I've got this working.

**Comment by tautschnig** [comment]
@thk123 Are you working off of a branch that has 3313f25a9fb5f667292d49bb57a359253ee19ffc?

**Comment by thk123** [comment]
@tautschnig thanks for the tip @smowton just pointed me towards that - now got the failure locally. It is to do with an invalid pop and only with symex driven lazy loading so will keep digging.

**Comment by thk123** [comment]
Tracked down the problem with the malformed pop (remove returns had already run on a function, so swapping out the type for the type in the symbol table meant no returns). Just checking locally that passes all the tests, then will push up here to verify on CI and then will create a TG bump since plausible it will break things there too.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 8366f3b5e47c11a4a1f2790057ba4b23f95ecae8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97821114
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by thk123** [comment]
Even when the compile errors are fixed, this fails in TG for non-equal types in symex - this looks to be to do with the custom stubbing - manual bump incoming.

File: src/goto-programs/remove_virtual_functions.cpp
Comment (smowton) [reply src/goto-programs/remove_virtual_functions.cpp]: Not so, that only checked that the types are either `type_eq` or are pointers

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (smowton) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:658]: It isn't `member_exprt`, but rather the nameless expression that describes a `getfield` or `setfield` operand produced by the Java bytecode parser

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:655]: This commit contains format fixups that ought to be squashed into a previous commit

**Review by smowton — APPROVED** [review]
Can't approve my own code of course, but I agree with @thk123 and @kroening's parts

**Comment by thk123** [comment]
TG bump looks to be passing - will address existing comments and tidy up, then should be ready for a final review from @tautschnig

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 050dbec33c530bcefd7325ea55d727c301951901).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98066474

**Comment by tautschnig** [comment]
@thk123 Just send a ping when you're done with cleanup work.

**Comment by tautschnig** [comment]
@thk123 Actually there isn't anything new that I hadn't looked at in the meantime, so just do the cleanup, merge, and update TG accordingly. Thanks!

**Comment by thk123** [comment]
Am now happy with this, so review away @tautschnig 
I would particularly like some feedback on the [`root_object`](https://github.com/diffblue/cbmc/pull/3822/files#diff-ac8338bc55f009ec0eefcd32d58578b5R26) method in the unit tests - it feels like this logic may (must?) exist elsewhere already.

**Comment by tautschnig** [comment]
@thk123 re `root_object`: there is `object_descriptor_exprt`, which can be passed an `exprt` via its `build` method and then you get a `.root_object()` - but I need to re-read your code to see whether this does exactly what you need.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:666]: I'd use `const struct_typet &object_type = ...` here.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (tautschnig) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:681]: Dumb question: what is it that assures us that we always need to look at the first component? I get that because of Java not permitting multiple inheritance if anything, it must be the first one. Is it because everything inherits at least from `java.lang.object` that we know that this first component must be a struct type?

File: src/goto-programs/remove_virtual_functions.cpp
Comment (tautschnig) [inline src/goto-programs/remove_virtual_functions.cpp:142]: Avoid using `make_typecast` (it's now deprecated), do `call_args[i] = typecast_exprt(call_args[i], need_type);` instead.

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (tautschnig) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Is that a debugging left-over?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (tautschnig) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp:25]: As said in the comment, there is `object_descriptor_exprt`. I think it may be necessary to extend `object_descriptor_exprt::build_object_descriptor_rec` with code handling paired `dereference_exprt/address_of_exprt`, but then it should do the trick for you. I believe all the code that this needs in addition is (some variant of):
```
else if(expr.id() == ID_dereference)
{
  const dereference_exprt &deref = to_dereference_expr(expr);
  if(deref.pointer().id() == ID_address_of)
  {
    const address_of_exprt &address_of = to_address_of_expr(deref.pointer());
    dest.object() = address_of.object();
    build_object_descriptor_rec(ns, address_of.object(), dest);
  }
}
```

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: I'd prefer not mixing our headers with STL includes.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:666]: This code is in fact simply moved - I'd prefer to follow this PR up with a clean up PR as this whole block I think can be replaced by `resolve_inherited_componentt`

**Comment by thk123** [comment]
> what is it that assures us that we always need to look at the first component?

It probably isn't a great assumption - it is _true_ because in `java_bytecode_convert_class` the base class component is always added first (line 370) and then additional fields (line 458). As with other comment - would prefer to address in follow up PR using `resolve_inherited_componentt`

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (thk123) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp:25]: Yes I saw the `object_descriptor_exprt` and saw that it didn't handle the ref/deref. In the interest of not wanting to break anything else that uses this, I will attempt this refactor in a follow up PR too if that is acceptable to you.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (smowton) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:681]: Yep, invariant of Java object layout: every class has at least one member, and the first one is either `@class_identifier` (java.lang.Object, or any stub class whose position in the class hierarchy is unknown) or its single superclass.

**Comment by thk123** [comment]
Seem to have done enough rebases to make clang-format go away...

**Comment by thk123** [comment]
@tautschnig addressed minor comments, larger changes I would like to address in a follow up PR. If that's OK with you, this looks good to merge once CI passes.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5a9bf1ad65350f5a0fe074bf84f730aaab5ba88d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98303689
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f0c520e8aec055afddbebc95457e70bdc7eb521a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98304937

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp
Comment (tautschnig) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:666]: Sounds good!

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (tautschnig) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp:25]: Sure!

Comment (tautschnig) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp:25]: I'll post a PR in a few minutes.

Comment (tautschnig) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp:25]: See #3898.

**Pull request merged**
